folding-schemes/src/folding/ova/nifs.rs

```diff
@@ -0,0 +1,498 @@
+/// This module contains the implementation of the Ova scheme NIFS (Non-Interactive Folding Scheme) as
+/// outlined in the protocol description doc: <https://hackmd.io/V4838nnlRKal9ZiTHiGYzw?both#Construction>
+/// authored by Benedikt Bünz.
+use ark_crypto_primitives::sponge::Absorb;
+use ark_ec::{CurveGroup, Group};
+use ark_std::One;
+use std::marker::PhantomData;
+
+use super::{CommittedInstance, Witness};
+use crate::arith::r1cs::R1CS;
+use crate::commitment::CommitmentScheme;
+use crate::transcript::Transcript;
+use crate::utils::vec::{hadamard, mat_vec_mul, vec_scalar_mul, vec_sub};
+use crate::Error;
+
+/// Implements all the operations executed by the Non-Interactive Folding Scheme described in the protocol
+/// spec by Bünz in the [original HackMD](https://hackmd.io/V4838nnlRKal9ZiTHiGYzw?both#Construction).
+/// `H` specifies whether the NIFS will use a blinding factor
+pub struct NIFS<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool = false> {
+    _c: PhantomData<C>,
+    _cp: PhantomData<CS>,
+}
+
+impl<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool> NIFS<C, CS, H>
+where
+    <C as Group>::ScalarField: Absorb,
+{
+    /// Computes the T parameter (Cross Terms) as in Nova.
+    /// The wrapper is only in place to facilitate the calling as we need
+    /// to reconstruct the `z`s being folded in order to compute T.
+    pub fn compute_T(
+        r1cs: &R1CS<C::ScalarField>,
+        w_i: &Witness<C>,
+        x_i: &[C::ScalarField],
+        W_i: &Witness<C>,
+        X_i: &[C::ScalarField],
+        mu: C::ScalarField,
+    ) -> Result<Vec<C::ScalarField>, Error> {
+        crate::folding::nova::nifs::NIFS::<C, CS, H>::compute_T(
+            r1cs,
+            C::ScalarField::one(),
+            mu,
+            &[vec![C::ScalarField::one()], x_i.to_vec(), w_i.w.to_vec()].concat(),
+            &[vec![mu], X_i.to_vec(), W_i.w.to_vec()].concat(),
+        )
```

Comment:
This would be at the expense of creating all these `z` vectors in the middle of the code. I think this function is done to "hide" all that boilerplate. 

Also, notice that I don't need to pass `mu2`. So I can save up operations.

Reply:
but you construct it anyways here in this method in order to call the Nova's compute_T. Anyways, as mentioned in the other comments I'm fine with the current state of the code in the PR and I think we can merge